### PR TITLE
Message Codec

### PR DESCRIPTION
**Pull request by gatesn**
This PR extracts IO as much as possible from message reading/writing.

* A `MessageEncoder` takes either Buffer, DType, or ArrayData and produces a `Vec<Buffer>`
* A `MessageDecoder` takes a `BytesMut` and either returns a message, or the number of additional bytes it needs.
* `(A)SyncMessageReader` wraps a `(Async)Read` trait to produce messages. Vice-versa for `(A)SyncMessageWriter`.
* `(A)SyncIPCReader` provides a way of turning a Read into an `Array(Iterator|Stream)`
* `Array(Iterator|Stream)IPC` provides a way to turn an `Array(Iterator|Stream)` into ipc bytes

Part of #1676

### CODE REVIEW

File: vortex-file/src/read/layouts/chunked.rs [hunk 1]
```diff
@@ -420,22 +421,25 @@ mod tests {
         cache: Arc<RwLock<LayoutMessageCache>>,
         scan: Scan,
     ) -> (ChunkedLayoutReader, ChunkedLayoutReader, Bytes, usize) {
-        let mut writer = MessageWriter::new(Vec::new());
```
Comment (gatesn) [inline vortex-file/src/read/layouts/chunked.rs:423]: This messy test code will be refactored when I get around to the layouts

File: vortex-ipc/src/messages/reader_async.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    pub struct AsyncMessageReader<R> {
+        #[pin]
+        read: R,
+        buffer: BytesMut,
+        decoder: MessageDecoder,
+        bytes_read: usize,
+    }
+}
+
+impl<R: AsyncRead + Unpin> AsyncMessageReader<R> {
```
Comment (a10y) [inline vortex-ipc/src/messages/reader_async.rs]: I don't think you need all the `Unpin` bounds, if you're accessing the reader thru projection?

File: vortex-ipc/src/messages/decoder.rs [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    header: Buffer,
+    buffers_length: usize,
+}
+
+struct ReadingBuffer {
+    length: usize,
+    length_with_padding: usize,
+}
+
+pub enum NextMessage {
```
Comment (a10y) [inline vortex-ipc/src/messages/decoder.rs]: This is effectively the same as `PollRead` from vortex-io, can we do anything to make the names a bit more similar? Or perhaps even do some reuse? It's confusing to see the same concept in multiple places in the codebase under different names

File: vortex-ipc/src/messages/reader_sync.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    pub fn read_message(&mut self) -> VortexResult<Option<DecoderMessage>> {
+        loop {
+            match self.decoder.read_next(&mut self.buffer)? {
+                NextMessage::Some(msg) => {
+                    return Ok(Some(msg));
+                }
+                NextMessage::NeedMore(nbytes) => {
+                    self.buffer.resize(nbytes, 0x00);
```
Comment (a10y) [inline vortex-ipc/src/messages/reader_sync.rs:35]: isn't zeroing the memory slower than the usual thing we do

```
buffer.set_len(nbytes);
```

File: vortex-ipc/src/messages/writer_async.rs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    pub async fn write_message(&mut self, message: EncoderMessage<'_>) -> VortexResult<()> {
+        for buffer in self.encoder.encode(message) {
+            self.write.write_all(&buffer).await?;
+        }
+        Ok(())
+    }
+
+    pub fn get_mut(&mut self) -> &mut W {
```
Comment (a10y) [inline vortex-ipc/src/messages/writer_async.rs]: should this be an `AsMut` instead similar to the AsRef impl?

File: vortex-ipc/src/messages/reader_sync.rs [hunk 6]
```diff
@@ -0,0 +1,45 @@
+use std::io::Read;
+
+use bytes::BytesMut;
+use vortex_error::VortexResult;
+
+use crate::messages::{DecoderMessage, MessageDecoder, NextMessage};
+
+pub struct SyncMessageReader<R> {
```
Comment (a10y) [inline vortex-ipc/src/messages/reader_sync.rs:9]: shall we impl `Iterator<DecoderMessage>` to mirror the `Stream<DecoderMessage>` that we impl for the async variant?

File: vortex-ipc/src/messages/reader_sync.rs (hunk 4 above)
Comment (a10y) [reply vortex-ipc/src/messages/reader_sync.rs:35]: hmm i suppose since it reuses the allocation this shouldn't happen often

File: vortex-ipc/src/messages/decoder.rs [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+impl MessageDecoder {
+    /// Attempt to read the next message from the bytes object.
+    /// If the message is incomplete, the function will return `NeedMore` with the _total_ number
+    /// of bytes needed. The next call to read_next _should_ provide at least this number of bytes.
+    pub fn read_next(&mut self, bytes: &mut BytesMut) -> VortexResult<NextMessage> {
+        match &self.state {
+            State::Length => {
+                if bytes.len() < 4 {
+                    return Ok(NextMessage::NeedMore(4));
```
Comment (a10y) [inline vortex-ipc/src/messages/decoder.rs]: This should be `4 - bytes.len()`, if the goal is that `NeedMore(x)` means "I need x more bytes of data to be read from the stream"

Comment (a10y) [reply vortex-ipc/src/messages/decoder.rs]: If we want the semantic to be something else we should document it. In general pub tuple structs should be documented b/c otherwise you need to read a lot of code to figure out what they mean

File: vortex-ipc/src/iterator.rs [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+pub struct ArrayIteratorIPCBytes {
+    inner: Box<dyn ArrayIterator + 'static>,
+    encoder: MessageEncoder,
+    buffers: Vec<Buffer>,
+}
+
+impl ArrayIteratorIPCBytes {
+    /// Collects the IPC bytes into a single `Vec<u8>` buffer.
+    pub fn collect_to_buffer(self) -> VortexResult<Vec<u8>> {
```
Comment (a10y) [inline vortex-ipc/src/iterator.rs]: can't this be `FromIterator`? Also the naming is weird, it's collecting to vec not Buffer

File: vortex-ipc/src/iterator.rs [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        let mut encoder = MessageEncoder::new(ALIGNMENT);
+        let buffers = encoder.encode(EncoderMessage::DType(self.dtype()));
+        ArrayIteratorIPCBytes {
+            inner: Box::new(self),
+            encoder,
+            buffers,
+        }
+    }
+
+    fn write_ipc<W: Write>(self, mut write: W) -> VortexResult<W>
```
Comment (a10y) [inline vortex-ipc/src/iterator.rs:97]: why is this here and not on `ArrayIteratorIPCBytes` directly?

Comment (gatesn) [reply vortex-ipc/src/iterator.rs:97]: ArrayIteratorIPCBytes is an `Iterator<Item=VortexResult<Buffer>>`, I guess that could be written into a target? But typically iterator structs are single-purpose.

File: vortex-ipc/src/iterator.rs (hunk 8 above)
Comment (gatesn) [reply vortex-ipc/src/iterator.rs]: Can't really do a TryFromiterator.

Agree the name is odd, can change, I just didn't want it to be confused with `collect_vec` which would return a `VortexResult<Vec<Buffer>>`

File: vortex-ipc/src/messages/decoder.rs (hunk 3 above)
Comment (gatesn) [reply vortex-ipc/src/messages/decoder.rs]: Oh weird, it looks like it's defined in `vortex-file`? Happy to use the same name. In Tokio this whole mechanism is known as a "codec": https://docs.rs/tokio-util/latest/tokio_util/codec/index.html#the-decoder-trait

But not sure it should be reusable tbh. LayoutReaders ask for more messages, not more bytes. So it's a similar pattern to lift IO out of a reader, but not much logic to be shared.

Will make the names align though

File: vortex-ipc/src/messages/decoder.rs (hunk 7 above)
Comment (gatesn) [reply vortex-ipc/src/messages/decoder.rs]: I can move the doc-string to the enum. But the semantic is the number of total bytes for the next successful read.

File: vortex-ipc/src/messages/reader_async.rs (hunk 2 above)
Comment (gatesn) [reply vortex-ipc/src/messages/reader_async.rs]: good point, will try to remove them

File: vortex-ipc/src/messages/reader_sync.rs (hunk 4 above)
Comment (gatesn) [reply vortex-ipc/src/messages/reader_sync.rs:35]: 🤷 if it shows up on profiles can fix. Otherwise less unsafe is better

File: vortex-ipc/src/iterator.rs [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+pub struct ArrayIteratorIPCBytes {
+    inner: Box<dyn ArrayIterator + 'static>,
+    encoder: MessageEncoder,
+    buffers: Vec<Buffer>,
+}
+
+impl ArrayIteratorIPCBytes {
+    /// Collects the IPC bytes into a single `Vec<u8>` buffer.
+    pub fn collect_to_buffer(self) -> VortexResult<Vec<u8>> {
+        let mut buffer = vec![];
```
Comment (a10y) [inline vortex-ipc/src/iterator.rs]: this is why this PR is failing miri. You can't write this data into an unaligned Vec<u8> and then attempt to use it to back a PrimitiveArray

File: vortex-ipc/src/messages/decoder.rs [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+        if !self.is_empty() {
+            // If the buffer isn't empty, we shouldn't try to copy it. It could be huge and this
+            // is just an optimization to avoid the copy in `split_to_aligned`.
+            return;
+        }
+
+        // Reserve up to the worst-cast alignment
+        self.reserve(additional + align);
+        let padding = self.as_ptr().align_offset(align);
+        unsafe { self.set_len(padding) };
```
Comment (a10y) [inline vortex-ipc/src/messages/decoder.rs:285]: don't you want to account for the case where len != 0 before this method is called?

File: vortex-ipc/src/messages/encoder.rs [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    }
+}
+
+impl MessageEncoder {
+    /// Create a new message encoder that pads each message and buffer with the given alignment.
+    ///
+    /// ## Panics
+    ///
+    /// Panics if `alignment` is greater than `u16::MAX`.
+    pub fn new(alignment: usize) -> Self {
```
Comment (a10y) [inline vortex-ipc/src/messages/encoder.rs:41]: should also panic if not a power of 2

File: vortex-ipc/src/messages/reader_sync.rs [hunk 13]
```diff
@@ -0,0 +1,51 @@
+use std::io::Read;
+
+use bytes::BytesMut;
+use vortex_error::VortexResult;
+
+use crate::messages::{DecoderMessage, MessageDecoder, PollRead};
+
+/// An IPC message reader backed by an `AsyncRead` stream.
```
Comment (a10y) [inline vortex-ipc/src/messages/reader_sync.rs]: backed by a std::io::Read

**Review by a10y — APPROVED** [review]
